File: Day4/day4.py
with open("input.txt") as f:
    part_one_count = 0
    part_two_count = 0

    for section_pair in f:
        section_one, section_two = section_pair.strip("\n").split(",")
        
        start_one, end_one = list(map(int, section_one.split("-")))
        start_two, end_two = list(map(int, section_two.split("-")))

        # To calculate number of pairs that encapsulate one another
        if (start_one >= start_two and end_one <= end_two) or (start_two >= start_one and end_two <= end_one):
            part_one_count += 1

        # To calculate number of pairs that overlap one another
        if (start_one <= end_two and start_two <= end_one) or (start_two <= end_one and start_one <= end_two):
            part_two_count += 1

    # Part 1 Answer
    print(part_one_count)

    # Part 2 Answer
    print(part_two_count)

File: Day04/day4.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class day4 {
    public static void main(String[] args) {

        int partOneAnswer = 0;
        int partTwoAnswer = 0;
        try {
            BufferedReader reader = new BufferedReader(new FileReader("input.txt"));

            String input;
            while ((input = reader.readLine()) != null) {
                String[] sections = input.split(",");
                String[] sectionOne = sections[0].split("-");
                String[] sectionTwo = sections[1].split("-");

                int startOne = Integer.parseInt(sectionOne[0]);
                int endOne = Integer.parseInt(sectionOne[1]);
                int startTwo = Integer.parseInt(sectionTwo[0]);
                int endTwo = Integer.parseInt(sectionTwo[1]);

                // To calculate number of pairs that encapsulate one another
                if ((startOne >= startTwo && endOne <= endTwo) || (startTwo >= startOne && endTwo <= endOne)) {
                    partOneAnswer++;
                }
        
                // To calculate number of pairs that overlap one another
                if ((startOne <= endTwo && startTwo <= endOne) || (startTwo <= endOne && startOne <= endTwo)) {
                    partTwoAnswer++;
                }
            }
            reader.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        // Part 1 Answer
        System.out.println(partOneAnswer);

        // Part 2 Answer
        System.out.println(partTwoAnswer);
    }
}
